draft new example for linked libs (#56)
runtimeverification/gitbook-kontrol: guides/kontrol-example/linked-library-example.md

There is one more branch that takes place.

We could inspect the kcfg once more to see what the control flow graph looks like now, but that is optional, as the information we need right now is also present in the final output of Kontrol.

<figure><img src="../../.gitbook/assets/linked_libs_5.png" alt=""><figcaption><p>Kontrol_TUI_2</p></figcaption></figure>

The output summarizes the first failure identified while running the proof.
We determine if a node represents a failure based on multiple conditions, such as a revert status code or a failed assertion.
[You can learn more about the success predicate here](https://github.com/runtimeverification/kontrol/blob/master/src/kontrol/kdist/foundry.md#foundry-success-predicate).
It has three components: the failure reason, the path condition, and the model.
The output shows that the status code is EVMC_REVERT.

Let's focus on the path condition for a bit.
We can see that it is a conjunction of two negated conditions:
    1. `#Not ( { VV0_n_114b9705:Int #Equals chop ( VV0_n_114b9705:Int *Int VV0_n_114b9705:Int ) /Word VV0_n_114b9705:Int } )`
    2. `#Not ( { VV0_n_114b9705:Int #Equals 0 } )`

> The second one is pretty straightforward, asserting that the symbolic argument `VV0_n_114b9705`, is non-zero.

Reviewer: ```suggestion
The second one is pretty straightforward, asserting that the symbolic argument `VV0_n_114b9705` (a K representation of `uint256 n`) is non-zero.
```

For the first one, we can simplify the variable name to obtain `#Not( n #Equals chop (n *Int n /Word n))`.
Let's break it apart:

> - `chop()` is an internal function that ensures that it's argument is lower than `2^256`.

Reviewer: ```suggestion
- `chop()` is an internal function that ensures that its argument is less than `2^256`.
```

Think about it as a (`_%_`) mod operation.

> - In the K language, operators also take the sort on which they are applied.

Reviewer: Not sure about this statement: do you mean that the result is of the same sort as arguments? Or that the operators should only be applied to a specific sort? Either way, the next sentence is very clear :)

  `*Int` represents the product of two unbounded integers.

> The equation above is equivalent to `( n != (n * n / n) mod type(uint256).max)` which looks more like an overflow check.

Reviewer: I think it might actually be 
```
( n != (n * n / n) mod (type(uint256).max + 1))
```
Because `type(uint256).max` is `2^256 - 1` (115792089237316195423570985008687907853269984665640564039457584007913129639935), and `chop` is `modInt pow256`.

Author: Good catch!

> And that's what it actually represents, it is a compiler inserted check for the operation inside the `square` function.

Reviewer: ```suggestion
And that's what it actually represents: a compiler-inserted check for the operation inside the `square` function.
```

> So, the proof will lead to an overflow that will throw a revert whenever n is not 0 and n is not equal to `((n * n / n) mod type(uint256).max)`.

Reviewer: _I think:_
```suggestion
In other words, the proof will run into an overflow and revert whenever `n` is not 0 and is not equal to `((n * n / n) mod (type(uint256).max) + 1)`.
```

> The model also gives as a counter-example value from `n`, under `VV0_n_114b9705 = pow128`.

Reviewer: ```suggestion
The model also demonstrates `VV0_n_114b9705 = pow128` as a counter-example--the value that  `n` should take for the failure to occur.
```

> In Solidity, to avoid overflow when squarring a number, the result must stay within the `uint256` range, which has a maximum value of `2^256 - 1`.

Reviewer: ```suggestion
In Solidity, to avoid overflow when squaring a number, the result must stay within the `uint256` range, which has a maximum value of `2^256 - 1`.
```
